File: examples/go/inductive-token/InductiveToken_test.go
package contract

import (
	"testing"
	runar "github.com/icellan/runar/packages/runar-go"
)

var (
	alice   = runar.PubKey("alice_pubkey_33bytes_placeholder!")
	bob     = runar.PubKey("bob___pubkey_33bytes_placeholder!")
	tokenId = runar.ByteString("test-token-001")
)

func newToken(owner runar.PubKey, balance runar.Bigint) *InductiveToken {
	return &InductiveToken{Owner: owner, Balance: balance, TokenId: tokenId}
}

func TestInductiveToken_Transfer(t *testing.T) {
	c := newToken(alice, 100)
	c.Transfer(runar.MockSig(), bob, 30, 1000)
	out := c.Outputs()
	if len(out) != 2 {
		t.Fatalf("expected 2 outputs, got %d", len(out))
	}
	if out[0].Values[0] != bob {
		t.Error("output[0] owner should be bob")
	}
	if out[0].Values[1] != runar.Bigint(30) {
		t.Errorf("output[0] balance: expected 30, got %v", out[0].Values[1])
	}
	if out[1].Values[0] != alice {
		t.Error("output[1] owner should be alice")
	}
	if out[1].Values[1] != runar.Bigint(70) {
		t.Errorf("output[1] balance: expected 70, got %v", out[1].Values[1])
	}
}

func TestInductiveToken_Transfer_ExactBalance(t *testing.T) {
	c := newToken(alice, 100)
	c.Transfer(runar.MockSig(), bob, 100, 1000)
	out := c.Outputs()
	if len(out) != 2 {
		t.Fatalf("expected 2 outputs, got %d", len(out))
	}
	if out[0].Values[1] != runar.Bigint(100) {
		t.Errorf("output[0] balance: expected 100, got %v", out[0].Values[1])
	}
	if out[1].Values[1] != runar.Bigint(0) {
		t.Errorf("output[1] balance: expected 0, got %v", out[1].Values[1])
	}
}

func TestInductiveToken_Transfer_ZeroAmount_Fails(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure")
		}
	}()
	newToken(alice, 100).Transfer(runar.MockSig(), bob, 0, 1000)
}

func TestInductiveToken_Transfer_ExceedsBalance_Fails(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure")
		}
	}()
	newToken(alice, 100).Transfer(runar.MockSig(), bob, 101, 1000)
}

func TestInductiveToken_Send(t *testing.T) {
	c := newToken(alice, 100)
	c.Send(runar.MockSig(), bob, 1000)
	out := c.Outputs()
	if len(out) != 1 {
		t.Fatalf("expected 1 output, got %d", len(out))
	}
	if out[0].Values[0] != bob {
		t.Error("output[0] owner should be bob")
	}
	if out[0].Values[1] != runar.Bigint(100) {
		t.Errorf("output[0] balance: expected 100, got %v", out[0].Values[1])
	}
}

func TestInductiveToken_Compile(t *testing.T) {
	if err := runar.CompileCheck("InductiveToken.runar.go"); err != nil {
		t.Fatalf("Runar compile check failed: %v", err)
	}
}

File: examples/rust/inductive-token/InductiveToken_test.rs
use runar::prelude::*;

#[derive(Clone)]
struct ItOutput { satoshis: Bigint, owner: PubKey, balance: Bigint }

struct InductiveToken {
    owner: PubKey,
    balance: Bigint,
    token_id: ByteString,
    outputs: Vec<ItOutput>,
}

impl InductiveToken {
    fn add_output(&mut self, satoshis: Bigint, owner: PubKey, balance: Bigint) {
        self.outputs.push(ItOutput { satoshis, owner, balance });
    }

    fn transfer(&mut self, sig: &Sig, to: PubKey, amount: Bigint, output_satoshis: Bigint) {
        assert!(check_sig(sig, &self.owner));
        assert!(amount > 0);
        assert!(amount <= self.balance);

        self.add_output(output_satoshis, to, amount);
        let change_owner = self.owner.clone();
        self.add_output(output_satoshis, change_owner, self.balance - amount);
    }

    fn send(&mut self, sig: &Sig, to: PubKey, output_satoshis: Bigint) {
        assert!(check_sig(sig, &self.owner));

        self.add_output(output_satoshis, to, self.balance);
    }
}

fn alice() -> PubKey { b"alice_pubkey_33bytes_placeholder!".to_vec() }
fn bob() -> PubKey { b"bob___pubkey_33bytes_placeholder!".to_vec() }

fn new_token(owner: PubKey, balance: Bigint) -> InductiveToken {
    InductiveToken { owner, balance, token_id: b"test-token-001".to_vec(), outputs: vec![] }
}

#[test]
fn test_transfer() {
    let mut c = new_token(alice(), 100);
    c.transfer(&mock_sig(), bob(), 30, 1000);
    assert_eq!(c.outputs.len(), 2);
    assert_eq!(c.outputs[0].owner, bob());
    assert_eq!(c.outputs[0].balance, 30);
    assert_eq!(c.outputs[1].owner, alice());
    assert_eq!(c.outputs[1].balance, 70);
}

#[test]
fn test_transfer_exact_balance() {
    let mut c = new_token(alice(), 100);
    c.transfer(&mock_sig(), bob(), 100, 1000);
    assert_eq!(c.outputs.len(), 2);
    assert_eq!(c.outputs[0].balance, 100);
    assert_eq!(c.outputs[1].balance, 0);
}

#[test]
#[should_panic]
fn test_transfer_zero_amount_fails() {
    new_token(alice(), 100).transfer(&mock_sig(), bob(), 0, 1000);
}

#[test]
#[should_panic]
fn test_transfer_exceeds_balance_fails() {
    new_token(alice(), 100).transfer(&mock_sig(), bob(), 101, 1000);
}

#[test]
fn test_send() {
    let mut c = new_token(alice(), 100);
    c.send(&mock_sig(), bob(), 1000);
    assert_eq!(c.outputs.len(), 1);
    assert_eq!(c.outputs[0].owner, bob());
    assert_eq!(c.outputs[0].balance, 100);
}

#[test]
fn test_compile() {
    runar::compile_check(
        include_str!("InductiveToken.runar.rs"),
        "InductiveToken.runar.rs",
    ).unwrap();
}
